### ISSUE TITLE
Mac OS X 64 memory protection fault during gc

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1478]
I haven't been able to reproduce this locally, but it is preventing build uploads.  

Ping @bjourne just in case you have an idea what it might be.

```
Unit Test: { [ ] [ anton's-regression ] }
Unit Test: { { t } [ ffi_test_61 bool-and-ptr? ] }
Unit Test: { { { t f } } [ ffi_test_61 [ b>> ] [ ptr>> ] bi 2array ] }
Unit Test: {
    { S{ uint-pair { a 2882400171 } { b 305419896 } } }
    [ ffi_test_62 ]
}
Unit Test: {
    {
        S{ ulonglong-pair
            { a 12379814471884843981 }
            { b 1311768467302729063 }
        }
    }
    [ ffi_test_63 ]
}
Unit Test: {
    { t }
    [
        10 [ [ 100 random ] twice 2array ] replicate [
            [
                first2 [ <alien> ] dip bug1021_test_1
                alien-address
            ] map
        ] [ [ first2 sq + ] map ] bi =
    ]
}
Unit Test: {
    { }
    [
        minor-gc 2000
        [ 101 <alien> run-test [ alien-address ] map drop ]
        times
    ]
}
Unit Test: { { } [ 10000 [ 0 doit 33 assert= ] times ] }
Loading resource:basis/compiler/tests/call-effect.factor
Must Fail With: { [ 1 f execute-ic-test ] [ second 3 = ] }
Unit Test: { [ ] [ [ ] call-test ] }
Unit Test: { [ ] [ f [ drop ] curry call-test ] }
Unit Test: { [ ] [ [ ] [ ] compose call-test ] }
Must Fail With: { [ [ 1 2 3 ] call-test ] [ wrong-values? ] }
Loading resource:basis/compiler/tests/callback-error.factor
Unit Test: {
    [ ]
    [
        " USING: alien alien.c-types alien.syntax kernel ;\n    IN: scr..."
        callback-error-script ascii set-file-contents
    ]
}
Unit Test: { [ t ] [ run-vm-with-script "\"Error!\"" swap member? ] }
Unit Test: {
    [ ]
    [
        "USING: alien alien.c-types alien.syntax kernel threads ;\n    ..."
        callback-error-script ascii set-file-contents
    ]
}
Unit Test: { [ t ] [ run-vm-with-script "\"Error!\"" swap member? ] }
Loading resource:basis/compiler/tests/callstack-overflow.factor
Unit Test: { { t } [ 10 [ recurse ] replicate [ ] all? ] }
Must Fail With: { [ overflow-c ] [ 2 head { "kernel-error" 16 } = ] }
Must Fail With: { [ overflow/w-primitive ] [ 2 head { "kernel-error" 16 } = ] }
Unit Test: { { } [ overflow/w-compact-gc ] }
fatal_error: Memory protection fault during gc: 0x100283ff8
```

**Comment by bjourne** [comment]
Yeah I planned to look at it. So it doesn't happen locally on your mac 64 bit but it happens on the build server? If so it is very weird.

**Comment by mrjbq7** [comment]
I'm going to try a ``load-all test-all`` again and see.  Will report back.

**Comment by bjourne** [comment]
I made a commit which should make the dump output more detailed when Factor runs into a fatal error. Like:

```
fatal_error: Im fatal!: 0

Nursery   : [0x7f1b6e180000 -> 0x7f1b6e380000]    2097152 bytes
Aging     : [0x7f1b6d980000 -> 0x7f1b6dd80000]    4194304 bytes
Tenured   : [0x7f1b61980000 -> 0x7f1b6d980000]  201326592 bytes
Cards     : [0x7f1b618a4010 -> 0x7f1b6196e410]     828416 bytes

Contexts:

  Active:
  Datastack   : [0x7f1b7823d000 -> 0x7f1b7827d000]     262144 bytes
...
```
But I can't see it in the build log here: http://builds.factorcode.org/report?os=macosx&cpu=x86.64 Is mason maybe truncating the crash log at the first line? If it was shown in full it would be easier to debug.

**Comment by erg** [comment]
I'm out but we should maybe use the without-limits combinator somewhere?

**Comment by mrjbq7** [comment]
That's dangerous of generating emails can fill up disk space or exceed email limits of not careful.

**Comment by mrjbq7** [comment]
Okay, I can reproduce this error.

Let me try with a debug build.

**Comment by mrjbq7** [comment]
Seems to pass with debug build??

**Comment by mrjbq7** [comment]
Okay, thats weird. Now it passes with release build.  Some kind of heisenbug.

**Comment by bjourne** [comment]
If you get it to crasch again, can you paste the dump? The fatal error says where in the memory the segfault happens and it should be close to the bottom of the active contexts callstack.

**Comment by mrjbq7** [comment]
@bjourne with latest code it doesn't give me any different callstack dump, could it be there are two places that throw fatal error and one you updated and another isn't using the new code dump on error?

**Comment by bjourne** [comment]
Maybe but I don't see it. There is only one `fatal_error` and it always calls `dump_memory_layout`. I'm out of ideas myself. Maybe we should disable the test for mac?

**Comment by mrjbq7** [comment]
It could be ``dump_memory_layout`` is just taking forever to run and I only see the first line printed out and kill the process after awhile.  Let me run it longer.

**Comment by mrjbq7** [comment]
Hmm, no ``factor`` is not using CPU, and not exiting, and only has ``fatal_error: Memory protection fault during gc: 0x100683ff8`` printed...

**Comment by bjourne** [comment]
Maybe because the callstack is kaputt at that point. Something interesting to try is to comment out all calls to `set_border_locked`. By default, memory is mmapped read-write  so that should make the unit test pass.

**Comment by mrjbq7** [comment]
Consistently reproducible on clean checkout:

```
$ factor
IN: scratchpad load-all save test-all
```

**Comment by bjourne** [comment]
Then I'm out of ideas. It's like the code that turns off the memory protection for the page doesn't have any effect.

**Comment by nicolas-p** [comment]
Is this the reason why the last OS X 64-bit build was in August?

![capture d ecran 2015-12-07 a 12 52 46](https://cloud.githubusercontent.com/assets/11431580/11626245/8f5ab17c-9ce1-11e5-8fb5-7de97f897df0.png)

**Comment by mrjbq7** [comment]
Yes although building from source still works fine it's just the tests trigger this issue (preventing a binary from being uploaded).  We will hopefully have a workaround/fix soon but have some holiday delays.

**Comment by erg** [comment]
https://github.com/slavapestov/factor/commit/b8aef640f9387697efdf747d028adcd9bcabd1bf

**Comment by mrjbq7** [comment]
Do you think @bjourne exposed a latent bug by adding page protection? I thought the pages already had mprotect and that was a refactor?

**Comment by erg** [comment]
Nah, looks like that patch is probably not the culprit, but it was that block of patches that exposed/caused the bug.

**Comment by bjourne** [comment]
Let's just disable the  test for mac 64bit? It appears to be working fine for 32bit macs.

**Comment by mrjbq7** [comment]
I don't think it's a single failure in a specific test, is it?  I only reproduced with a `load-all test-all`, and consistent failure but not sure an obvious test to ignore?

**Comment by bjourne** [comment]
Yes it is the test `Unit Test: { { } [ overflow/w-compact-gc ] }` that always fails. Currently mac 32bit also fails but with an unrelated error.

**Comment by mrjbq7** [comment]
Ah, maybe you can disable so we get new binaries and then have an issue to fix it.

**Comment by erg** [comment]
Some info!

The compact-gc call doesn't have enough callstack space and crashes. It crashes in the ``RT_DLSYM`` case:

```c++
cell factor_vm::lookup_external_address(relocation_type rel_type,
                                        code_block *compiled,
                                        array* parameters,
                                        cell index) {
  switch (rel_type) {
    case RT_DLSYM:
      return compute_dlsym_address(parameters, index, false);
```

``compact-gc`` shouldn't be doing a call to dlsym in the first place, but just increasing the callstack space from 4k to 16k stops the crash. However, this has the disadvantage of bloating each coroutine (green thread) by 16k, so it's not the ideal fix.

The best fixes would be to stop calling dlsym and to use a separate coroutine to call vm functions.

Additionally, there's mention of a callstack size of 16384 in ``basis/compiler/tests/callstack-overflow.factor``, is that still valid? Maybe we can reenable the test on Windows after fixing things.
